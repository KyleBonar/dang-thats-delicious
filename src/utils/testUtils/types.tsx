export interface MockLink {
  text: string;
  to: string;
  target: "_blank" | "_self";
}
