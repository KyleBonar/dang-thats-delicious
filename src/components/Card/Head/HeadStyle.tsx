import styled from "../../../theme/styled-components";
import { Link } from "../../Link/Link";

export const StyledImage = styled.img`
  width: 100%;
`;
StyledImage.displayName = "img";

export const StyledLink = styled(Link)`
  button {
    text-transform: uppercase;
    font-size: 14px;
    letter-spacing: 1px;
    margin: auto auto 0;
  }
`;
StyledLink.displayName = "Link";
