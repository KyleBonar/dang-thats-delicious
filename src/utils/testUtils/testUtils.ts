import * as React from "react";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";

import { MockCard } from "./types";
import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import { FlashMessageProps } from "../../components/FlashMessage/FlashMessage";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.string,
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.array_of_words(),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = new Array(casual.integer(1, 15))
    .fill(null)
    .map(fakeSauce);
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured: undefined,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const storeConfig = configureStore([]);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const fakeCard = (): MockCard => ({
  title: casual.string,
  description: casual.string,
  to: casual.url,
  showLink: casual.boolean,
  imageLink: casual.url,
  className: casual.string,
  anchorText: casual.string
});

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(0, MIN_PASSWORD_LENGTH);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword
};
