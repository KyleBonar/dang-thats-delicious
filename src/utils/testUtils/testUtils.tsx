import * as React from "react";
import * as enzyme from "enzyme";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";
import { IErrReturn } from "../Err/Err";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.random_element([undefined, casual.string]),
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.random_element([undefined, casual.array_of_words()]),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const generateFakeSauces = (): ISauce[] =>
  new Array(casual.integer(1, 15)).fill(null).map(fakeSauce);

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = generateFakeSauces();
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  const featured: undefined | string[] = casual.random_element([
    undefined,
    allSlugs.length > 0
      ? allSlugs.filter(slug => {
          if (casual.random < 0.4) return slug;
          return;
        })
      : undefined
  ]);

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const middlewares = [thunk];
  const storeConfig = configureStore(middlewares);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(1, MIN_PASSWORD_LENGTH - 1);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

// Function designed to resemble how a network error would look like.
// This will help the catch(err) only have to handle a single format
const generateErr = (): IErrReturn => {
  return {
    response: {
      data: {
        status: 400,
        msg: casual.string,
        isGood: false
      }
    }
  };
};

// lets us mount up a simple custom hook
const mountReactHook = hook => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.shallow(
      <Component>
        {hookValues => {
          Object.assign(componentHook, hookValues);
          return null;
        }}
      </Component>
    );
  });
  return { componentMount, componentHook };
};

export interface ImountReactHookWithReduxStore {
  componentMount: any;
  componentHook: any;
}
const mountReactHookWithReduxStore = (
  hook,
  store
): ImountReactHookWithReduxStore => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.mount(
      <Provider store={store}>
        <Component>
          {hookValues => {
            Object.assign(componentHook, hookValues);
            return null;
          }}
        </Component>
      </Provider>
    );
  });
  return { componentMount, componentHook };
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword,
  generateInValidPassword,
  generateErr,
  mountReactHook,
  mountReactHookWithReduxStore
};
