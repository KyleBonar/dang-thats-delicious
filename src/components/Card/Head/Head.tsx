import * as React from "react";

import { StyledLink, StyledImage } from "./HeadStyle";

interface IHeadProps {
  showLink?: boolean;
  to: string;
  imageLink?: string;
  description?: string;
}

const Head: React.FC<IHeadProps> = props => {
  const { showLink, imageLink, description, to } = props;

  if (!showLink) return <StyledImage src={imageLink} alt={description} />;

  return (
    <StyledLink href={to}>
      <StyledImage
        src={imageLink}
        alt={description}
        onError={e => {
          const elem = e.target as HTMLImageElement;
          elem.src =
            "https://res.cloudinary.com/foryourthoughts/image/upload/v1575869743/sauces/ra1o7bsr9v2eurosoo5y_bktfsa.png";
        }}
      />
    </StyledLink>
  );
};

export default Head;
