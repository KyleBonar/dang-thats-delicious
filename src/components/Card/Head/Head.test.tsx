import * as React from "react";
import * as enzyme from "enzyme";

import Head, { IHeadProps } from "./Head";
import { casual, ITERATION_SIZE } from "../../../utils/testUtils/testUtils";

const fakeHead = (): IHeadProps => ({
  showLink: casual.random_element([undefined, casual.boolean]),
  to: casual.string,
  imageLink: casual.random_element([undefined, casual.string]),
  description: casual.random_element([undefined, casual.string])
});

const mockHead = new Array(ITERATION_SIZE).fill(null).map(fakeHead);

describe("<Head />", () => {
  it("renders", () => {
    const wrapper = enzyme.shallow(<Head to="" />);

    expect(wrapper).toBeTruthy();
  });

  it("matches snapshot", () => {
    const wrapper = enzyme.shallow(<Head to="" />);

    expect(wrapper).toMatchSnapshot();
  });

  it("if showLink is false, do not render Link component", () => {
    const wrapper = enzyme.shallow(<Head to="" />);

    expect(wrapper.exists("Link")).toBeFalsy();
  });
});
