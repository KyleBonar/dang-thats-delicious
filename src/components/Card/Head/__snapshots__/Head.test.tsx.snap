// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Head /> matches snapshot 1`] = `
<Link
  href="corporis sed veritatis quidem praesentium magnam fugiat"
>
  <img
    onError={[Function]}
  />
</Link>
`;

exports[`<Head /> matches snapshot 2`] = `
<Link
  href="maiores recusandae fuga reprehenderit soluta non omnis"
>
  <img
    onError={[Function]}
  />
</Link>
`;

exports[`<Head /> matches snapshot 3`] = `
<img
  alt="perferendis dolorum cupiditate illum saepe ut assumenda"
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 4`] = `
<Link
  href="exercitationem omnis laborum exercitationem est sapiente laudantium"
>
  <img
    onError={[Function]}
    src="assumenda quasi adipisci mollitia iste itaque omnis"
  />
</Link>
`;

exports[`<Head /> matches snapshot 5`] = `
<img
  onError={[Function]}
  src="et non explicabo dolores ipsam laboriosam non"
/>
`;

exports[`<Head /> matches snapshot 6`] = `
<img
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 7`] = `
<img
  alt="at sint dolores fugiat quis et voluptas"
  onError={[Function]}
  src="in sunt occaecati nulla aut sunt est"
/>
`;

exports[`<Head /> matches snapshot 8`] = `
<img
  alt="sint officiis qui dignissimos dolore qui voluptatem"
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 9`] = `
<img
  alt="autem quam qui facere in ut odit"
  onError={[Function]}
  src="quisquam aut consequatur sint provident aut ullam"
/>
`;

exports[`<Head /> matches snapshot 10`] = `
<Link
  href="occaecati at odit tempora at modi occaecati"
>
  <img
    alt="consequatur itaque et in fugit in rerum"
    onError={[Function]}
    src="laborum quos eum voluptate sit saepe assumenda"
  />
</Link>
`;

exports[`<Head /> matches snapshot 11`] = `
<img
  alt="libero aut autem ut ullam et possimus"
  onError={[Function]}
  src="ut ducimus fugiat et et aut quidem"
/>
`;

exports[`<Head /> matches snapshot 12`] = `
<img
  alt="et optio consequatur velit voluptate necessitatibus excepturi"
  onError={[Function]}
  src="minus quo sunt possimus earum animi quis"
/>
`;

exports[`<Head /> matches snapshot 13`] = `
<Link
  href="nesciunt provident eaque atque quos rerum nobis"
>
  <img
    onError={[Function]}
  />
</Link>
`;

exports[`<Head /> matches snapshot 14`] = `
<Link
  href="accusantium molestiae non sint et ipsa consequuntur"
>
  <img
    alt="quis quis nobis repudiandae aliquam molestiae occaecati"
    onError={[Function]}
    src="dolorem quia consequatur omnis ab assumenda animi"
  />
</Link>
`;

exports[`<Head /> matches snapshot 15`] = `
<img
  alt="quia aut sint cumque sunt mollitia quibusdam"
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 16`] = `
<img
  onError={[Function]}
  src="est labore sed rerum nesciunt voluptas rem"
/>
`;

exports[`<Head /> matches snapshot 17`] = `
<img
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 18`] = `
<img
  alt="voluptas eos quam consequatur ipsam sed nobis"
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 19`] = `
<img
  alt="occaecati eveniet aut eligendi iusto sed pariatur"
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 20`] = `
<img
  alt="enim rem qui hic dolore quibusdam quasi"
  onError={[Function]}
  src="quis ratione eveniet ea totam aut aut"
/>
`;

exports[`<Head /> matches snapshot 21`] = `
<Link
  href="explicabo at quae autem eligendi nihil quis"
>
  <img
    onError={[Function]}
    src="sit voluptas dicta non dolore autem aut"
  />
</Link>
`;

exports[`<Head /> matches snapshot 22`] = `
<img
  onError={[Function]}
  src="laboriosam tempore voluptatibus reiciendis ut eos commodi"
/>
`;

exports[`<Head /> matches snapshot 23`] = `
<img
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 24`] = `
<img
  alt="accusamus molestiae quia est reprehenderit commodi perspiciatis"
  onError={[Function]}
  src="quasi distinctio reiciendis perspiciatis ut illo sit"
/>
`;

exports[`<Head /> matches snapshot 25`] = `
<img
  alt="aut dolorem in non dolorem nihil optio"
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 26`] = `
<img
  alt="consequatur provident qui delectus qui qui ab"
  onError={[Function]}
  src="impedit dolores saepe dolor incidunt vero quia"
/>
`;

exports[`<Head /> matches snapshot 27`] = `
<img
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 28`] = `
<img
  alt="officiis sed illum et ea quia sunt"
  onError={[Function]}
  src="et voluptas repellendus nisi corrupti quas eligendi"
/>
`;

exports[`<Head /> matches snapshot 29`] = `
<img
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 30`] = `
<img
  alt="perferendis soluta dignissimos cum non ea sed"
  onError={[Function]}
  src="non nulla sit officia doloremque non sit"
/>
`;

exports[`<Head /> matches snapshot 31`] = `
<img
  onError={[Function]}
/>
`;

exports[`<Head /> matches snapshot 32`] = `
<img
  alt="odit in eos sunt ipsum saepe veritatis"
  onError={[Function]}
  src="inventore vitae sint eligendi quis excepturi veritatis"
/>
`;
